src/Fantomas/Context.fs

```diff
@@ -149,6 +190,28 @@ let internal dumpAndContinue (ctx: Context) =
 #endif
     ctx
 
+let internal isShortExpression maxWidth (shortExpression: Context -> Context) (fallbackExpression) (ctx: Context) =
+    // create special context that will process the writer events slightly different
+    let shortExpressionContext = ctx.WithShortExpression(maxWidth)
+    let resultContext = shortExpression shortExpressionContext
+    match resultContext.WriterModel.Mode with
+    | ShortExpression info ->
+        // verify the expression is not longer than allowed
+        if info.ConfirmedMultiline
+        then fallbackExpression ctx
+        else
+            let shortExpressionEvents =
+                Seq.skip info.PreviousEventCount resultContext.WriterEvents
+                |> Queue.ofSeq
+            let events =
+                Queue.fold (fun acc event -> Queue.conj event acc) ctx.WriterEvents shortExpressionEvents
+
+            let model = WriterModel.updateAll shortExpressionEvents ctx.WriterModel
+            { ctx with WriterEvents = events; WriterModel = model }
```

Comment:
(potential perf problem)
unless I'm missing something this could be replaced by
```suggestion
            { resultContext with WriterModel = { resultContext.WriterModel with Mode = ctx.WriterModel.Mode } }
```

if that's not a case, then this should be quicker (`Queue.rev` is almost free here) :
```suggestion
            let shortExpressionEvents =
                resultContext.WriterEvents |> Queue.rev |> Seq.take (Queue.length resultContext.WriterEvents - info.PreviousEventCount) 
                |> Queue.ofSeq |> Queue.rev
            let events =
                Queue.fold (fun acc event -> Queue.conj event acc) ctx.WriterEvents shortExpressionEvents

            let model = WriterModel.updateAll shortExpressionEvents ctx.WriterModel
            { ctx with WriterEvents = events; WriterModel = model }
```

Reply:
Good thinking, that is indeed a cheaper solution.